rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1059 @@
+else
+  go (dirOf path);
+```
+
+**Rationale and alternatives**
+
+- Attribute sets and lists could start on the same line as the if keywords, saving an indentation level on their body:
+  ```nix  
+  #1a
+  if builtins.length matches != 0 then {
+    inherit path matches;
+  } else if path == /. then [
+    1
+    2
+  ] else
+    go (dirOf path);
+  ```
+  - This results in inconsistent vertical start of the keywords, making the structure harder to follow
+- Have the `then` on the start of the next line, directly followed by the if body:
+  ```nix
+  #1b
+  if builtins.length matches != 0
+    then { inherit path matches; }
+  else if path == /.
+    then [
+      1
+      2
+    ]
+  else go (dirOf path);
+
+  #1c
+  if builtins.length matches != 0
+  then { inherit path matches; }
+  else if path == /.
+  then [
+    1
+    2
+  ]
+  else go (dirOf path);
+  ```
+
+### with, assert
+
+**Description**
+
+- The body after the statement starts on a new line, without indentation.
+- There may be exceptions for common idioms, in which the body already starts on the same line.
+
+**Examples**
+
+```nix
+with pkgs;
+assert foo == bar; {
+  meta.maintainers =
+    with lib.maintainers; [
+      some
+      people
+    ];
+}
+```
+
+### let
+
+**Description**
+
+- Let bindings are *always* multiline.
+- The "let" part is indented one level, but not the "in" part.
+- Each item in the "let" part is indented and starts on its own line.
+  For more details, see the [binders section](#binders).
+- The "in" part starts on a new line.
+
+**Examples**
+
+```nix
+let
+  foo = "bar";
+in
+if foo == "bar" then
+  "hello"
+else
+  "world"
+```
+
+### Attribute sets and lists
+
+**Description**
+
+- Brackets and braces are generally written with a space on the inside, like `[ `, ` ]`, `{ ` and ` }`.
+  - Empty lists and attribute sets are written as`[ ]` and`{ }`, respectively.
+- Lists and attribute sets with multiple items are liberally expanded.
+  - They can only be on a single line if:
+    - They contain at most one element
+    - Fit on the line
+  - As described under [Binders](#binders) below, nested attribute sets are always expanded.
+
+**Examples**
+
+```nix
+#1
+[
+  { }
+  { foo = "bar"; }
+  {
+    foo = {
+      bar = "baz";
+    };
+  }
+  { foo.bar = "baz"; }
+]
+
+#2
+[
+  [ 1 ]
+  [
+    2
+    3
+  ]
+]
+
+#3
+[
+  [
+    1
+    2
+    3
+  ]
+]
+
+#4
+[
+  {
+    mySingletons = [
+      [
+        ({
+          # stuff in there
+        })
+      ]
+    ];
+  
+    mySingletons' = [
+      [
+        (function call)
+      ]
+    ];
+  }
+]
+```
+
+**Drawbacks**
+
+- Singleton lists may use a lot of indentation
+
+**Rationale and alternatives**
+
+- Have a special compact form for singleton lists, to reduce the indentation level and remove two additional lines
+  ```nix
+  foo = [ {
+    # content
+  } ];
+  ```
+- Be a bit less eager about expanding lists and attribute sets (for example by allowing up to n elements on a single line).
```

Comment:
Yes, and the following heuristic could work: ~5 non-nested elements (average human short-term memory) that fit on a line.

Reply:
Maybe this decision should be moved out of the RFC. I don't think it matters whether it's 1 or 5, just needs to be some number, and we shouldn't be afraid to change it if it turns out a different one works better.

Reply:
We now changed this section to allow "few" elements on the same line, without giving a specific number: https://github.com/NixOS/rfcs/pull/166/commits/385a914e9a506b6b24506a770d71f29e3cd65fdb

Reply:
You call it flexible, but to me, this looks like an inconsistency. I hope not.

Reply:
Without inconsistency, we either have
- cluttered diffs (when expanding lists less eagerly) or
- wasted space, eyesight clutter (if we expand lists more eagerly).

With inconsistency, the human can decide whether a specific list should be expanded or not (up to a point where the list length hits the line length limit).

However, I also really like the premise of the same code having the same output formatting, no matter the input formatting.

Reply:
This happens because alejandra has no line length limit.

Reply:
> kamadorueda/alejandra#373

This is not a bug, but a feature, that I would expect this RFC to provide as well. As long as the line limit allows, the formatter should respect the author's decision whether a list / attrset should remain on a single line or not. A common case when humans make better decisions that formatters is when you have two or more list/attrset bindings which are strongly related. In such cases, alignment can make the code easier for the human reader. There are various tools that automate this, like https://github.com/junegunn/vim-easy-align

> However, the output format should be as pure a function of the AST as possible, and thus as independent of the input format as possible.

Why? Are we optimizing the output for humans or for machines? If we want simplicity and consistency, we can just print everything on one line. The Nix evaluator won't mind. I hope we agree that this is not what we want. Optimizing for diff readability is optimizing for humans, so it inherentlymakes the formating non-consistent. We should embrace that, instead of fighting our nature, if we want satisfying results. Of course, consistency is important, but we should not take it to extremes (e.g. printing everything one line).

>> want to be able to write code (or to emit generated code) onto a single line with full confidence that the formatter will do its job correctly.

> This happens because alejandra has no line length limit.

Yes, indeed, it's a question of whether a particular formatter supports line limit or not. In general I find Prettier's output to be one of the best across all the formatters I've used for any programming language.
In their [rationale](https://prettier.io/docs/en/rationale.html)
they claim that they consider this a workaround and not a feature, but after having using both Prettier and Alejandra for a long time I find this a good design tradeoff, as it ultimate works very satisfyingly in practice.

Reply:
> > However, the output format should be as pure a function of the AST as possible, and thus as independent of the input format as possible.
>
> Why? Are we optimizing the output for humans or for machines? If we want simplicity and consistency, we can just print everything on one line. The Nix evaluator won't mind. I hope we agree that this is not what we want. Optimizing for diff readability is optimizing for humans, so it inherentlymakes the formating non-consistent. We should embrace that, instead of fighting our nature, if we want satisfying results. Of course, consistency is important, but we should not take it to extremes (e.g. printing everything one line).

This. I 100% agree. Trying to format the code based solely on the AST is an anti-pattern and I don't really understand, why do some people insist on it. Stuff like newlines, empty lines and comments are not represented in the AST, but are extremely important for good formatting. Instead of operating on the AST, formatters should work with the CST which includes sufficient information to perfectly reconstruct arbitrary source code (including all the non-functional stuff like whitespace and comments).

Reply:
I think, that it's important to recognize that humans are going to try to optimize how things look anyway. Even if we make the formatter "a pure function of the AST (+comments? +empty lines?)", that still won't prevent people from adjusting their code so that it looks better when formatted. Stuff like inserting parentheses, comments, empty lines, extracting intermediate expressions into `let` blocks, etc etc.

My reading of the `prevent future debate around how things should be formatted` goal is that we want to avoid bikeshedding about code formatting in PRs while still keeping a generally uniform style across nixpkgs. In other words, running the formatter should automatically produce code that is "good enough" to pass the PR smell check, not that the tool should completely eliminate the possibility of having two equivalent pieces of code that are formatted differently.

IMHO, within the space of "good enough to get the PR accepted" there is some space for manual adjustment by the author (who is a human and therefore may be able to format code better than just "good enough"). Rules like "keep empty lines for logical grouping", "manually decide whether to split a small list or to keep it on a single line", etc seem pretty reasonable and I don't expect them to lead to much "`future debate around how things should be formatted`". If a PR reviewer is nitpicky enough to block the PR due to such formatting choices, no amount of "pure deterministic AST formatting" is going to stop them from bikeshedding.

Reply:
A primary goal of this RFC is to not bother humans with the format of the code, *at all*. All stated goals of this RFC directly derive from this, although I admit that it should be written down explicitly as well.

There are two main issues with respecting formatting choices of the input code:

1. The *possibility* to make things slightly prettier will always result in an *expectation* to do so (even if it's only an internally motivated one, I know my perfectionist friends), weakening the main point of this RFC. Those attempts at tweaking the output might also lead to "fighting the auto formatter", which is frustrating and wasteful.
2. The formatter has no ability do distinguish between "I just vomited all my code onto one line because it's not my job to format it" and "I deliberately put things onto one line because I think it looks better". Therefore, preserving aspects of the input format automatically means that *everyone* has to pre-format their code, or else it will look significantly worse. This is not acceptable.

Yes humans can produce a better and prettier output than a tool. But the goal is to have the format *good enough*, and nobody should waste their time trying to make it slightly better.

---------

@RuRo
Obviously nixfmt currently uses some form of CST, or else it wouldn't be preserving comments and empty lines. I prefer think of it as an "annotated AST", but technically it matches the CST definition. Anyways, details.

I think that preserving empty lines and nothing else is a good compromise, as doing so does not exhibit most of the drawbacks I described above.

- Most code looks good enough without anyways, so nobody is forced to add them to have their code look good
- They are a crucial tool to group items together, especially given that the formatter destroys all other kinds of grouping (in a list, every item must go onto its own line, you can't do "one line per group")
- They do not really affect whether or not a syntax element will be expanded or not, because you can only really add them into already expanded code

--------

> Even if we make the formatter "a pure function of the AST (+comments? +empty lines?)", that still won't prevent people from adjusting their code so that it looks better when formatted. Stuff like inserting parentheses, comments, empty lines, extracting intermediate expressions into let blocks, etc etc.

If people are improving their code structure, that's always a win. This is also where the "Bad code does not deserve good formatting." mantra comes from: We don't want to try to make shitty code bearable (at least beyond a certain point; it's a gradient), we want to encourage people to make their code less shitty in the first place. But this is more a secondary effect than a goal on its own.

Reply:
An example on how one could currently fight the formatter:
```nix
{
  foo = [
    5
    14
    6
    23
    1
    54
  ];
  bar = [5] ++ [14] ++ [6] ++ [23] ++ [1] ++ [54];
}

```

Reply:
I think that having formatting directives in the code could be an acceptable compromise here. They do allow customization for those who really want to, but it doesn't cause the main issue of leaving the formatter at a guess about whether a certain input format was "intentional" or not.

I'd like to note though that all of this is essentially an implementation detail based off nixfmt's current design, and that it has little to no influence over the format described in this RFC: This RFC mostly specifies the layout of individual syntax elements when expanded onto multiple lines, on most accounts it (by design) remains rather vague about the rules of when to expand a syntax element or not. I'd consider any such changes as "minor" which could thus easily be tweaked afterwards if necessary.

Reply:
Adding comments can also work as a way to fight the formatter:
```nix
{
  foobar = lib.loooongfoobarfun (bar: foo: eggspam (foosomething foo) (barsomething bar)) (
    builtins.readDir ./foobar
  );
  fmtbar =
    lib.loooongfoobarfun # wrap!
      (bar: foo: eggspam (foosomething foo) (barsomething bar))
      (builtins.readDir ./foobar);
}

```

Reply:
Discussed this in the meeting today. We think that some "fights" between the user and the formatter are unavoidable, but this is more of an implementation detail and doesn't need to be written down in the RFC.

Regarding the idea of allowing users to influence the formatter _just_ from the input layout, the shepherds think that it's not worth doing that compared to the current proposal as stated from the arguments in this thread.